sway/input/keyboard.c

```diff
@@ -494,6 +514,52 @@ static void handle_keyboard_group_key(struct wl_listener *listener,
 	handle_key_event(sway_group->seat_device->keyboard, data);
 }
 
+static void handle_keyboard_group_enter(struct wl_listener *listener,
+		void *data) {
+	struct sway_keyboard_group *sway_group =
+		wl_container_of(listener, sway_group, enter);
+	struct sway_keyboard *keyboard = sway_group->seat_device->keyboard;
+	struct wl_array *keycodes = data;
+
+	uint32_t *keycode;
+	wl_array_for_each(keycode, keycodes) {
+		struct key_info keyinfo;
+		update_keyboard_state(keyboard, *keycode, WLR_KEY_PRESSED, &keyinfo);
+	}
+}
+
+static void handle_keyboard_group_leave(struct wl_listener *listener,
+		void *data) {
+	struct sway_keyboard_group *sway_group =
+		wl_container_of(listener, sway_group, leave);
+	struct sway_keyboard *keyboard = sway_group->seat_device->keyboard;
+	struct wl_array *keycodes = data;
+
+	uint32_t *keycode;
+	wl_array_for_each(keycode, keycodes) {
+		struct key_info keyinfo;
+		update_keyboard_state(keyboard, *keycode, WLR_KEY_RELEASED, &keyinfo);
+
+		bool pressed_sent = update_shortcut_state(&keyboard->state_pressed_sent,
+				*keycode, WLR_KEY_RELEASED, keyinfo.keycode, 0);
+		if (!pressed_sent) {
+			continue;
+		}
+
+		struct wlr_seat *wlr_seat = keyboard->seat_device->sway_seat->wlr_seat;
+		struct wlr_input_device *wlr_device =
+			keyboard->seat_device->input_device->wlr_device;
+
+		struct timespec now;
+		clock_gettime(CLOCK_MONOTONIC, &now);
+		int64_t time_msec = (int64_t)now.tv_sec * 1000 + now.tv_nsec / 1000000;
+
+		wlr_seat_set_keyboard(wlr_seat, wlr_device);
+		wlr_seat_keyboard_notify_key(
```

Comment:
We need to properly update the clients' state too, without causing any "real" key press/release which would trigger an in-app shortcut for instance.

Reply:
I think the only way to do this would be a wl_keyboard leave immediately followed by an enter. I think this could be done once and only if any of the keys had the pressed event sent. I'll switch it over to do that tomorrow.

Reply:
The focused surface is now being refocused when at least one keycode had a pressed event sent. This should update the surface and prevent in app bindings from being triggered.